### PR TITLE
feat(composer): Use COMPOSER_AUTH var instead of file creation

### LINKED ISSUE #6907: Reconsider Composer auth.json
Like `.npmrc` for npm, adding the `auth.json` file to a repository can cause potential conflicts or undesired behavior. There seem to be [alternatives](https://getcomposer.org/doc/articles/handling-private-packages-with-satis.md#authentication).

Possible:
- `composer config http-basic.example.org username password` (also available with `--global`)
- Write `auth.json` to home directory
- [`COMPOSER_AUTH`](https://getcomposer.org/doc/03-cli.md#composer-auth) environment variable

I think env normally works best for us.

Cc @viceice @zharinov @danez - what do you think?

### PR DESCRIPTION
**Pull request by zharinov**
Closes #6907

### CODE REVIEW

**Comment by zharinov** [comment]
Haven't tested it on real repos

File: lib/manager/composer/artifacts.ts [hunk 1]
```diff
@@ -20,7 +20,68 @@ import {
...
+  }
+  try {
+    // istanbul ignore else
+    if (is.array(config.registryUrls)) {
+      for (const regUrl of config.registryUrls) {
+        if (regUrl) {
+          const { host } = URL.parse(regUrl);
+          const hostRule = hostRules.find({
+            hostType: datasourcePackagist.id,
+            url: regUrl,
```
Comment (rarkins) [inline lib/manager/composer/artifacts.ts:59]: Is there any downside if we just do a lookup without specifying `url` and use all `hosttype=packagist` authentication?

Comment (zharinov) [reply lib/manager/composer/artifacts.ts:59]: If I understand it right, in this case we need to handle multiple `HostRule` fields, e.g. `hostName`, `domainName` and `baseUrl`.

Comment (zharinov) [reply lib/manager/composer/artifacts.ts]: Take a look at changed code which you can find good enough (`baseUrl` handling requires additional `try-catch` and logging).

Comment (rarkins) [reply lib/manager/composer/artifacts.ts]: registryUrls might be empty or incomplete. I don't think we want to look up based on it. Let's have this change do GitHub / gitlab as we already do and then add others after.

File: lib/manager/composer/artifacts.ts [hunk 2]
```diff
@@ -22,6 +20,42 @@ import { getRepoStatus } from '../../util/git';
...
+    url: 'https://gitlab.com/api/v4/',
+  });
+  if (gitlabCredentials?.token) {
+    authJson['gitlab-token'] = {
+      'gitlab.com': gitlabCredentials.token,
+    };
+  }
+
+  hostRules
+    .findAll({ hostType: datasourcePackagist.id })
```
Comment (danez) [inline lib/manager/composer/artifacts.ts:47]: I tried to add hostRules for a git repo in `renovate.json`, but as we only consider credentials of `hostType: packagist` here it never arrived in composer. I wonder if it makes sense to loosen this check a little, by either removing the filter at all or also adding hostrules where `hostType == null`?

This is what I tried in `renovate.json`:
```
  "hostRules": [
    {
      "domainName": "bitbucketserver",
      "username": "composer",
      "password": "xxx"
    }
  ]
```

**Review by danez — COMMENTED** [review]
I tested this by changing `.devcontainer/Dockerfile` and installing `php-cli` with apt-get there and copying composer into the container (`COPY --from=composer:latest /usr/bin/composer /usr/bin/composer`)

Then I did 3 tests against a test repo (see https://gist.github.com/danez/5190fb0e57e8e1609f76579011478d2f#file-composer-json)

1. Not auth.json commited

- Renovate fails, because composer is not able to clone the git repo. Same error as with current version of renovate which overwrites auth.json (see https://gist.github.com/danez/5190fb0e57e8e1609f76579011478d2f#file-run-without-auth-json-log )

2. With commited auth.json

- Works as expected (see https://gist.github.com/danez/5190fb0e57e8e1609f76579011478d2f#file-run-with-auth-json-log)

3. Not auth.json commited, but credentials in renovate.json

- This did not work. I tried adding hostRules for our git repo, but it is not supplied to composer (see inline comment)

So the bug I tried to fix in #5653 is completely covered in this PR 🎊

File: lib/manager/composer/artifacts.ts (hunk 2 above)
Comment (viceice) [reply lib/manager/composer/artifacts.ts:47]: @danez You need to add `hostType: packagist` to your package rule

Comment (danez) [reply lib/manager/composer/artifacts.ts:47]: Okay

**Review by danez — APPROVED** [review]

**Pull request closed without merging**